SLEEF is faster than MKL for many vml ops

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #91962
* #91963
* #91961
* #91967
* #91960

vml appears to be an order of magnitude slower than `vec::map` on some
operators. For example, `sqrt` with numel=1e6 takes 339 us on master and 13.3 us
with this PR for a 25x speedup. I was able to reproduce on both intel and AMD
CPUs.

peterbell10 committed:
Don't use vml on AMD cpus

intel famously degrades performance on non-intel hardware, so vml is an order of
magnitude slower here than vec::map. For `sqrt` with numel=1e6 I see 339 us
on master and 13.3 us with this PR for a 25x speedup.

[ghstack-poisoned]

peterbell10 committed:
Update on "Don't use vml on AMD cpus"


intel famously degrades performance on non-intel hardware, so vml is an order of
magnitude slower here than vec::map. For `sqrt` with numel=1e6 I see 339 us
on master and 13.3 us with this PR for a 25x speedup.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

peterbell10 committed:
Update on "Don't use vml on AMD cpus"


intel famously degrades performance on non-intel hardware, so vml is an order of
magnitude slower here than vec::map. For `sqrt` with numel=1e6 I see 339 us
on master and 13.3 us with this PR for a 25x speedup.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

sanchitintel commented:
> intel famously degrades performance on non-intel hardware

Intel's optimizations may be unique to its products' microarchitecture. Intel does not guarantee the availability, functionality, or effectiveness of any optimization on microprocessors not manufactured by Intel.

cc @jgong5

peterbell10 committed:
Update on "SLEEF is faster than MKL for many vml ops"


vml appears to be an order of magnitude slower than `vec::map` on some
operators. For example, `sqrt` with numel=1e6 takes 339 us on master and 13.3 us
with this PR for a 25x speedup. I was able to reproduce on both intel and AMD
CPUs.

[ghstack-poisoned]

peterbell10 commented:
I've testing with an intel CPU and it looks like I was jumping to conclusions. With `vec::map`, `a.sqrt()` is roughly 2x slower than `a.clone()` but with a conda install of PyTorch on an intel machine I see `a.sqrt()` as 25x slower than `a.clone()`. Maybe there is a bug in vml, or the way we are using vml?

peterbell10 committed:
Update on "SLEEF is faster than MKL for many vml ops"


vml appears to be an order of magnitude slower than `vec::map` on some
operators. For example, `sqrt` with numel=1e6 takes 339 us on master and 13.3 us
with this PR for a 25x speedup. I was able to reproduce on both intel and AMD
CPUs.

[ghstack-poisoned]

peterbell10 commented:
After looking at `a.sqrt()` under perf I noticed that because I used `randn` to generate the random input, most of the time is spent in the error handling code. Sleef is only faster here because it propagates the NaN without any further checks. When I only pass positive numbers, the two roughly on par.

I tried using the mode flags to disable error handling but the performance doesn't improve much. I suppose nans shouldn't be coming up too often so I'll just close this.
